### PR TITLE
[web] Prevent crash when restore is called incorrectly after recording ends.

### PR DESCRIPTION
**Pull request by ferhatb**
## Description

This PR makes web consistent with mobile platforms.

When PaintingContext is used with a painter and component such as TextField, the painter itself causes paintContext.canvas to change to a new instance since one recording ends and another one starts. The code in ClipContext currently is calling save and restore on different Canvas(s) causing an exception on web since it has extra checks.

## Related Issues

Fixed https://github.com/flutter/flutter/issues/61697

## Tests

Updated recording_canvas_test to cover case.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/surface/recording_canvas.dart [hunk 1]
```diff
@@ -115,7 +120,7 @@ class RecordingCanvas {
   void endRecording() {
     _pictureBounds = _paintBounds.computeBounds();
     if (assertionsEnabled) {
-      _debugRecordingEnded = true;
+      _recordingEnded = true;
     }
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/surface/recording_canvas.dart:124]: The check for `assertionsEnabled` should be removed, right?

File: lib/web_ui/test/engine/recording_canvas_test.dart [hunk 2]
```diff
@@ -187,6 +187,14 @@ void main() {
     expect(mockCanvas.methodCallLog[2].methodName, 'restore');
     expect(mockCanvas.methodCallLog[3].methodName, 'endOfPaint');
   });
+
+  // Regression test for https://github.com/flutter/flutter/issues/61697.
+  test('Allows restore calls after recording has ended', () {
+    final RecordingCanvas rc = RecordingCanvas(Rect.fromLTRB(0, 0, 200, 400));
+    rc.endRecording();
+    // Should not throw exception on restore.
+    rc.restore();
```
Comment (mdebbar) [inline lib/web_ui/test/engine/recording_canvas_test.dart:196]: Using `expect` will probably show a nicer error on failure:
```dart
expect(() => rc.restore(), returnsNormally);
```

Comment (ferhatb) [reply lib/web_ui/test/engine/recording_canvas_test.dart:196]: Done.

File: lib/web_ui/lib/src/engine/surface/recording_canvas.dart (hunk 1 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/surface/recording_canvas.dart:124]: Thanks for spotting that!. Done.

File: lib/web_ui/test/engine/recording_canvas_test.dart [hunk 3]
```diff
@@ -187,6 +187,14 @@ void main() {
     expect(mockCanvas.methodCallLog[2].methodName, 'restore');
     expect(mockCanvas.methodCallLog[3].methodName, 'endOfPaint');
   });
+
```
Comment (nturgut) [inline lib/web_ui/test/engine/recording_canvas_test.dart:190]: very nit: doc comment on the method.

**Pull request merged**